lib/Sema/DerivedConformanceAdditiveArithmetic.cpp

```diff
@@ -54,38 +54,230 @@ static StringRef getMathOperatorName(MathOperator op) {
+      auto *newDecl = new (C) VarDecl(/*IsStatic*/ false, oldDecl->getIntroducer(),
+                                      /*IsCaptureList*/ false, oldDecl->getNameLoc(),
+                                      oldDecl->getName(), oldDecl->getDeclContext());
+      newDecl->setHasNonPatternBindingInit();
+      newDecl->setImplicit();
+      return newDecl;
+    };
+    llvm::transform(llvm::concat<VarDecl *>(lhsPayloadVars, rhsPayloadVars),
+                    std::begin(caseBodyVarDecls),
+                    copyDecl);
+
+    return CaseStmt::create(C, CaseParentKind::Switch, SourceLoc(),
+                            labelItem, SourceLoc(), SourceLoc(), caseBody, caseBodyVarDecls);
+  };
+  llvm::transform(enumDecl->getAllElements(),
+                  std::back_inserter(cases),
+                  createCaseForEnumElement);
+
+  // Create 'default: fatalError()' case for mismatching enum values
+  // (only necessary if the enum has more than one element)
+  if (cases.size() > 1) {
+    auto defaultPattern = AnyPattern::createImplicit(C);
+    auto defaultItem = CaseLabelItem::getDefault(defaultPattern);
+    auto fatalErrorDecl = getFatalErrorFn(C);
+    auto fatalErrorDRE = new (C) DeclRefExpr(fatalErrorDecl, DeclNameLoc(),
+                                             /*implicit*/ true);
+    auto errorMsg = new (C) StringLiteralExpr("Enum cases mismatch", SourceRange(),
+                                              /*implicit*/ true);
+    auto callExpr = CallExpr::createImplicit(C, fatalErrorDRE, {errorMsg}, {});
+    auto body = BraceStmt::create(C, SourceLoc(), ASTNode(callExpr),
+                                  SourceLoc());
+    cases.push_back(CaseStmt::create(C, CaseParentKind::Switch, SourceLoc(),
+                    defaultItem, SourceLoc(), SourceLoc(),
+                    body, /*case body var decls*/ None));
+  }
+
+  // Create 'switch (lhs, rhs) { <case statements> }'
+  auto lhsRef = new (C) DeclRefExpr(lhs, DeclNameLoc(), /*implicit*/true);
+  auto rhsRef = new (C) DeclRefExpr(rhs, DeclNameLoc(), /*implicit*/true);
+  auto lhsRhsExpr = TupleExpr::create(C, SourceLoc(), { lhsRef, rhsRef }, {}, {},
+                                      SourceLoc(), /*hasTrailingClosure*/ false,
+                                      /*implicit*/ true);
+  auto switchStmt = SwitchStmt::create(LabeledStmtInfo(), SourceLoc(), lhsRhsExpr,
+                                      SourceLoc(), cases, SourceLoc(), C);
+
+  auto body = BraceStmt::create(C, SourceLoc(), ASTNode(switchStmt), SourceLoc());
+  return { body, false };
 }
 
 // Synthesize body for math operator.
 static std::pair<BraceStmt *, bool>
-deriveBodyMathOperator(AbstractFunctionDecl *funcDecl, MathOperator op) {
+deriveBodyMathOperator_struct(AbstractFunctionDecl *funcDecl, MathOperator op) {
   auto *parentDC = funcDecl->getParent();
   auto *nominal = parentDC->getSelfNominalTypeDecl();
+  assert(isa<StructDecl>(nominal) && "Must be a struct");
```

Comment:
It would be nice to use the convenience functions (and rename the variable) for clarity:
```suggestion
  auto *structDecl = parentDC->getSelfStructDecl();
  assert(structDecl && "Must be a struct");
```